packages/dds/tree/src/simple-tree/api/identifierIndex.ts

```diff
@@ -0,0 +1,270 @@
+			: (schemaIdentifier: TreeNodeSchemaIdentifier) => {
+					const schemus = indexableSchemaMap.get(schemaIdentifier);
+					if (schemus !== undefined) {
+						const keyLocation = indexer(schemus);
+						if (keyLocation !== undefined) {
+							return makeGenericKeyFinder<TKey>(brand(keyLocation), isKeyValid);
+						}
+					}
+				};
+
+	const index = new AnchorTreeIndex<TKey, TValue>(
+		(view as SchematizingSimpleTreeView<TFieldSchema>).getView().checkout.forest,
+		schemaIndexer,
+		(anchorNodes) => {
+			const simpleTreeNodes: TreeNode[] = [];
+			for (const a of anchorNodes) {
+				const simpleTree = getOrCreateSimpleTree(view, a);
+				if (!isTreeValue(simpleTree)) {
+					simpleTreeNodes.push(simpleTree);
+				}
+			}
+
+			if (hasElement(simpleTreeNodes)) {
+				return getValue(simpleTreeNodes);
+			}
+		},
+		(anchorNode: AnchorNode) => {
+			const simpleTree = getOrCreateSimpleTree(view, anchorNode);
+			if (!isTreeValue(simpleTree)) {
+				return treeNodeApi.status(simpleTree);
+			}
+		},
+	);
+
+	// all the type checking guarantees that we put nodes of the correct type in the index
+	// but it's not captured in the type system
+	return index as SimpleTreeIndex<TKey, TValue>;
+}
+
+/**
+ * An index that returns tree nodes given their associated identifiers.
+ *
+ * @alpha
+ */
+export type IdentifierIndex = SimpleTreeIndex<string, TreeNode>;
+
+function isStringKey(key: TreeIndexKey): key is string {
+	return typeof key === "string";
+}
+
+/**
+ * Creates an {@link IdentifierIndex} for a given {@link TreeView}.
+ *
+ * @alpha
+ */
+export function createIdentifierIndex<TSchema extends ImplicitFieldSchema>(
+	view: TreeView<TSchema>,
+): IdentifierIndex {
+	// For each node schema, find which field key the identifier field is under.
+	// This can be done easily because identifiers are their own field kinds.
+	const identifierFields = new Map<string, string>();
+	walkFieldSchema(view.schema, {
+		node: (schemus) => {
+			if (schemus instanceof ObjectNodeSchema) {
+				for (const fieldKey of schemus.fields.keys()) {
+					identifierFields.set(schemus.identifier, fieldKey);
+					break;
+				}
+			}
+		},
+	});
+
+	return createSimpleTreeIndex(
+		view,
+		(schemus) => identifierFields.get(schemus.identifier),
+		(nodes) => {
+			if (nodes.length > 1) {
+				throw new UsageError(
+					"cannot retrieve node from index: there are multiple nodes with the same identifier",
+				);
+			}
+
+			return nodes[0];
+		},
+		isStringKey,
+	);
+}
+
+/**
+ * Gets a simple tree from an anchor node
+ */
+function getOrCreateSimpleTree<TSchema extends ImplicitFieldSchema>(
+	view: TreeView<TSchema>,
+	anchorNode: AnchorNode,
+): TreeNode | TreeIndexKey {
+	return (
+		anchorNode.slots.get(proxySlot) ??
+		makeTreeNode((view as SchematizingSimpleTreeView<TSchema>).getView().context, anchorNode)
+	);
+}
+
+/**
+ * Make a tree node from an anchor node
+ */
+function makeTreeNode(context: Context, anchorNode: AnchorNode): TreeNode | TreeIndexKey {
```

Comment:
We already have code to do this as part of TreeNodeKernel.getOrCreateInnerNode.

We however do not seem to have a way to actually use it to implement getOrCreateSimpleTree.

I think the cleanest solution would be to modify src/simple-tree/core/treeNodeKernel.ts (which already has the  questionable makeTree import and a comment about it) to factor out a non-exported `flexNodeFromAnchor` function.

Then add an exported `treeNodeFromAnchor` which can be implemented by calling `flexNodeFromAnchor` then `getTreeNodeSchemaFromHydratedFlexNode`

Reply:
Is `getTreeNodeSchemaFromHydratedFlexNode` the correct function to call? It seems like that gets me the `TreeNodeSchema` but from there, how do I get the `TreeNode`?

Reply:
Never mind, I figured it out :)